### ISSUE TITLE
Parsing big file is slow

### ISSUE THREAD

**Issue by cuixiasister** [antlr/antlr4#4160]
The target language I use is C++, which is very slow when parsing large files, and the memory usage exceeds 90%. How can I improve the parsing speed

**Comment by jimidle** [comment]
You will need to provide more information. What files are you parsing? What grammar are you using? How big is a 'large' file, etc.

**Comment by cuixiasister** [comment]
I'm parsing STIL files with ANTLR, and parsing gets really slow when the file size is over 100M，We usually parse files that are around 100 megabits in size，Here is an example of a STIL file，In the Pattern module, V can consist of millions of lines
#

STIL 1.0;
Signals 
{    
   DIR In;    
   OE_ In;    
}
Timing "hello tester timing" 
{    
    WaveformTable one 
	{     
    	Period '500ns';   
		Waveforms 
		{
		     DIR    { 01  { '0ns' ForceDown/ForceUp; }}
		   OE_    { 01  { '0ns' ForceUp; '200ns' ForceDown/ForceUp;  '300ns' ForceUp; }}
		}    
	}           // end WaveformTable one
}               // end

Pattern "hello tester pattern" 
{      
    W one;      
    V { ALL=00; }      
    V { ALL=00; }      
}

**Comment by cuixiasister** [comment]
> You will need to provide more information. What files are you parsing? What grammar are you using? How big is a 'large' file, etc.
I'm parsing STIL files with ANTLR, and parsing gets really slow when the file size is over 100M，We usually parse files that are around 100 megabits in size，Here is an example of a STIL file，In the Pattern module, V can consist of millions of lines

STIL 1.0;
Signals
{
DIR In;
OE_ In;
}
Timing "hello tester timing"
{
WaveformTable one
{
Period '500ns';
Waveforms
{
DIR { 01 { '0ns' ForceDown/ForceUp; }}
OE_ { 01 { '0ns' ForceUp; '200ns' ForceDown/ForceUp; '300ns' ForceUp; }}
}
} // end WaveformTable one
} // end

Pattern "hello tester pattern"
{
W one;
V { ALL=00; }
V { ALL=00; }
}

**Comment by ericvergnaud** [comment]
Hi
This 'smells' like a memory shortage issue, where most time is spent swapping virtual memory.

**Comment by jimidle** [comment]
> 

Also, is this behavior shown when there is no code other than that generated by ANTLR? I.E. User code should be eliminated.

Maybe with this much input, the internal ANTLR runtime Cache needs to be reset occasionally? 100M isn't huge for modern memory, but all sorts of objects are probably accumulating.

**Comment by KvanTTT** [comment]
> I'm parsing STIL files with ANTLR, and parsing gets really slow when the file size is over 100M，We usually parse files that are around 100 megabits in size，Here is an example of a STIL file，In the Pattern module, V can consist of millions of lines

Yes, it's possible because ANTLR creates parse tree instead of AST. It causes high memory consumption and low performance. You can try using an approach with creating AST on the fly using listerners but it requires additional implementation. Also, take a look at this example: [PlSqlAntlrListenerConverter](https://github.com/KvanTTT/PT.PM/blob/dev/Sources/PT.PM.PlSqlParseTreeUst/PlSqlAntlrListenerConverter.cs).
